### ISSUE TITLE
Securing routes by JWT Claims

### ISSUE THREAD

**Issue by ashwini-desai** [micronaut-projects/micronaut-security#256]
It is a common practice in claim based authorization approach, to allow or disallow a user from accessing a resource, based on claims available in JWT. 

I see we have @RolesAllowed annotation that lets us do the same for role based authorization. Similar to [Issue#254](https://github.com/micronaut-projects/micronaut-security/issues/254), a SecurityRule could be provided for checking claims. This will help the requests to be rejected even before entering controller logic. E.g ASP.NET Core provides [@Authorize(...)](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies?view=aspnetcore-3.1)

The logic could be made generic as well since roles, scopes everything are a claim against which we want to implement all these checks.

Is something of this sort already in pipeline?

**Comment by graemerocher** [comment]
Not that I am aware of, maybe @jameskleeh can provide a better answer

PRs are of course welcome

**Comment by jameskleeh** [comment]
@ashwini-desai Can you provide an example of what claim you want to check? How would the framework know:

1. What claim to check
2. What value to compare it against
3. How to compare the values

Is this specific to scope? If so, this would duplicate https://github.com/micronaut-projects/micronaut-security/issues/254

**Comment by traycho** [comment]
I think this topic becomes more popular, i was working on something similar. 

@jameskleeh  Is it possible to incorporate something of this kind https://github.com/traycho/pulsarix-micronaut-security-jwt under `micronaut-security-jwt`

**Comment by ashwini-desai** [comment]
@jameskleeh Missed this thread earlier.

Example could be very similar to how @RolesAllowed is implemented. 
To answer your questions:

1. What claim to check - It could support a key value check like @ClaimAllowed("sub", "xyz") or like how we can configure a key for roles, we could let users defined custom key which could be an array etc that holds custom claims in a token and @ClaimsAllowed("value1", "value2") etc

2. What value to compare it against - Value can be provided in the annotation as mentioned above

3. How to compare the values - For key and value, it could be done easily, for @ClaimsAllowed where only values are passed, it could be done using claims key configured in security configuration.

**Comment by dstepanov** [comment]
You can override `RolesFinder` and provide extra roles from claims.

**Comment by ashwini-desai** [comment]
@dstepanov `RolesFinder` by the name is meant for checking `Roles` only. We could provide an implementation, but doesn't sound correct. Also, in general there could be a lot of rules between different claims in the token which could be provided as rules, which will be helpful. So just overriding `RolesFinder` might not help in the longer run. Check [this](https://gist.github.com/jcassee/8ce917387263d40b9908506af5b5c578) project as an example.

**Comment by sdelamo** [comment]
Yes, I think this is a new feature. I think we could support a `@ScopesAllowed` annotation or a similar name.

**Comment by ashwini-desai** [comment]
@sdelamo Yes, we could provide generic annotation which can take one or many security attributes from the token and expected values, give enhancement options for AND/OR between them. This will make it work for all attributes, we won't have to make specific annotations like `@ScopesAllowed` then.

**Comment by adrnola** [comment]
@sdelamo / @jameskleeh - Can you provide an update on this issue [256](https://github.com/micronaut-projects/micronaut-security/issues/256) which relates to [254](https://github.com/micronaut-projects/micronaut-security/issues/254). Do you foresee this being implemented anytime soon ? This would be especially useful for the [OAUTH 2.0 Credentials Flow (Machine-To-Machine)](https://auth0.com/docs/flows/client-credentials-flow) which is becoming more prevalent in usage

**Comment by jameskleeh** [comment]
@adrnola I'm not convinced its feasible to implement claims validation at the framework level because there is no sensible default on how to verify the value of the claim is correct. We could implement a simple `@ClaimEquals(claim, value)` sort of annotation but I'm not sure how valuable that is. We could also implement an annotation to assert one of the provided scopes is set in the claims, however each of these tasks is not very complex and can be achieved today in any given application. The API to implement this functionality is public and open to anyone to use. I can't give a timeline when we might implement something related to this feature, however if you have a proposal we are willing to evaluate it.

**Comment by adrnola** [comment]
@jameskleeh - Thanks for the prompt response. 

Your right, Security is paramount in any application. Why not have the Micronaut framework have the "best" OAUTH 2.0 Developer user story. Why not have a "fully" consistent OAUTH 2.0 implementation for validating JWT Claims ? Rather than each developer rolling their own. I'm just thinking of the DRY principle here. Wouldn't this be another win for Micronaut over it's competitors.

This is how I do it at the moment, as an example:

[1] application.xml (I have introduced some "extra" my-company-XXXX variables within the micronaut.security.token.jwt namespace which I reference in my CustomJwtClaimsValidator)
<code>
  security:
    enabled: true
    token:
      jwt:
        my-company-issuer: https://my-company-appName-test.auth0.com/
        my-company-audience: https://appName-api-test.dp.clp.axa.com
        my-company-environment-proxy: false
        my-company-environment-proxy-type: HTTP
        my-company-environment-proxy-address: my.company.proxy:8080
        signatures:
          jwks:
            auth0:
              url: https://my-company-appName-test.auth0.com/.well-known/jwks.json
</code>

[2] Define a CustomJwtClaimsValidator implements GenericJwtClaimsValidator to define how I want to custom validate the JWT Token. I use the application.xml to be able to switch AUTH0 environments based on my-company-issuer and my-company-audience variables, example, going from Test Environment to Production Environment.

This worked fine, while I had only 1 JWT Claim to validate, I was able to do it in CustomJwtClaimsValidator . Now as we expand, we have different JWT Claims for different Routes.

So something like @ClaimEquals in conjunction with new variables my-company-issuer and my-company-audience in the application.xml file and can pass in 1 or more scopes, where Micronaut would inject in the JWTClaimsSet to ClaimEquals something like:
@JwtAuthorize(flow= "client-credentials", claim="scope", values={"scopeValue1", "scopeValue2"}, matcher="All/Any"), if it's just a single value than matcher is not needed. Depending on the flow, it would check for different things against entries in the application.xml (like my-company-flow: client-credentials) or Micronaut JwtAuthorize implementation.

I know GenericJwtClaimsValidator is deprecated and I will need to move to something like [using SecurityRules](https://blog.wick.technology/micronaut-security-rule/)

Hope my use case above makes sense. 

Thanks again for taking the time to review.


Here's my CustomJwtClaimsValidator to show you what I am trying to achieve.

<code>
/**
 * Validate Auth0 JWT.
 */
@Flogger
@Singleton
@Requires(property = JwtClaimsValidator.PREFIX + ".my-company-myApp-jwt", notEquals = StringUtils.FALSE)
public class CustomJwtClaimsValidator implements GenericJwtClaimsValidator {

    private static final String CLAIM_SCOPE = "foo:bar";
    private static final String TOKEN_CLAIM_SCOPE = "scope";
    private static final String TOKEN_CLAIM_GRANT_TYPE = "gty";
    private static final String TOKEN_CLAIM_GRANT_TYPE_CREDENTIALS = "client-credentials";
    private static final String TOKEN_CLAIM_AUTHORIZED_PARTY = "azp";

    @Value("${micronaut.security.token.jwt.my-company-issuer'}")
    private String myCompanyIssuer;

    @Value("${micronaut.security.token.jwt.my-company-audience'}")
    private String myCompanyAudience;

    /**
     * @param claimsSet JWT Claims
     * @return True if the JWT is valid
     */
    public boolean validate(@NonNull JWTClaimsSet claimsSet) {

        boolean isValid = false;

        log.atFine().log("Validating JWT");

        try {

            boolean isIssuerValid = false;
            final String issuer = claimsSet.getIssuer();
            if (issuer != null && issuer.equals(myCompanyIssuer)) {
                isIssuerValid = true;
            }

            boolean isAudienceValid = false;
            final String audience = claimsSet.getAudience().get(0);
            if (audience != null && audience.equals(myCompanyAudience)) {
                isAudienceValid = true;
            }

            boolean isClaimValid = false;
            final String claim = claimsSet.getStringClaim(TOKEN_CLAIM_SCOPE);
            if ( claim != null && claim.equals(CLAIM_SCOPE)) {
                isClaimValid = true;
            }

            boolean isGrantTypeValid = false;
            final String grantTypeClaim = claimsSet.getStringClaim(TOKEN_CLAIM_GRANT_TYPE);
            if (grantTypeClaim != null && grantTypeClaim.equals(TOKEN_CLAIM_GRANT_TYPE_CREDENTIALS)) {
                isGrantTypeValid = true;
            }

            boolean isAuthorizedPartyValid = false;
            final String authorizedPartyClaim = claimsSet.getStringClaim(TOKEN_CLAIM_AUTHORIZED_PARTY);
            if (authorizedPartyClaim != null) {
                isAuthorizedPartyValid = true;
            }

            boolean isNotExpired = false;
            final Date expTime = claimsSet.getExpirationTime();
            if (expTime != null) {
                final Date now = new Date();
                if (!expTime.before(now)) {
                    isNotExpired = true;
                }
            }

            isValid = isIssuerValid && isAudienceValid && isClaimValid && isGrantTypeValid && isAuthorizedPartyValid && isNotExpired;

            if (isValid) {
                log.atFine().log("Validating JWT - Success");
            } else {
                log.atWarning().log("Validating JWT - Failed: expectedAudience: %s, isIssuerValid: %s, isAudienceValid: %s, isClaimValid: %s, isGrantTypeValid: %s, isAuthorizedPartyValid: %s, isNotExpired: %s",
                        myCompanyAudience, isIssuerValid, isAudienceValid, isClaimValid, isGrantTypeValid, isAuthorizedPartyValid, isNotExpired);
                log.atWarning().log("Validating JWT - Failed: JWT ClaimsSet: %s", claimsSet.toJSONObject());
            }

        } catch (ParseException e) {
            log.atSevere().withCause(e).log("Error Parsing Claim");
        }

        return isValid;
    }

    @Override
    public boolean validate(JwtClaims claims) {
        return validate(JWTClaimsSetUtils.jwtClaimsSetFromClaims(claims));
    }
}
</code>

**Comment by adrnola** [comment]
@jameskleeh - Thanks. I did some more investigating into this.

I now have the following configuration:

[1] CustomJwtClaimsValidator implements GenericJwtClaimsValidator - which I use for Authentication of JWT (isIssuerValid && isAudienceValid && isGrantTypeValid && isAuthorizedPartyValid && isNotExpired) by using ammplication.xml entries

[2] I created a Custom SecurityRule and created a corresponding Class/Method Annotation - which I use for Authorization (which checks value of the scope).

[3] applciation.xml
  security:
    enabled: true
    token:
      jwt:
        signatures:
          jwks:
            auth0:
              url: https://axa-clp-dp-pricing-test.eu.auth0.com/.well-known/jwks.json
              # These entries are not Micronaut Framework Specific - They are Project Specific.
              # They are placed here as they are JWT/Auth0 specific.
              issuer: https://myApp-test.eu.auth0.com/
              audience: https://myApp-test.dp.clp.axa.com
              grant-type: client-credentials

which worked a treat and is very flexible in my Controller:
@JwtAuthorization(scope = "foo:bar")
@Post(value = "{something}")
public HttpResponse<AppResponse> doSomething()

I am just wondering what is best practice here, move Authentication from CustomJwtClaimsValidator to SecurityRule and use SecurityRule to do both Authentication & Authorization ?

I think where mine and other people (on this thread) confusion arouse is that the Micronaut Doc's does not have a concrete example in the use of SecurityRule & Annotation to validate the JWT.

Thanks.

**Comment by adrnola** [comment]
@jameskleeh - Thanks. I did some more investigating into this.

I now have the following configuration:

[1] CustomJwtClaimsValidator implements GenericJwtClaimsValidator - which I use for Authentication of JWT (isIssuerValid && isAudienceValid && isGrantTypeValid && isAuthorizedPartyValid && isNotExpired) by using ammplication.xml entries

[2] I created a Custom SecurityRule and created a corresponding Class/Method Annotation - which I use for Authorization (which checks value of the scope).

[3] applciation.xml
  security:
    enabled: true
    token:
      jwt:
        signatures:
          jwks:
            auth0:
              url: https://myApp-test.auth0.com/.well-known/jwks.json
              # These entries are not Micronaut Framework Specific - They are Project Specific.
              # They are placed here as they are JWT/Auth0 specific.
              issuer: https://myApp-test.eu.auth0.com/
              audience: https://myApp-test.com
              grant-type: client-credentials

which worked a treat and is very flexible in my Controller:
@JwtAuthorization(scope = "foo:bar")
@Post(value = "{something}")
public HttpResponse<AppResponse> doSomething()

I am just wondering what is best practice here, move Authentication from CustomJwtClaimsValidator to SecurityRule and use SecurityRule to do both Authentication & Authorization ?

I think where mine and other people (on this thread) confusion arouse is that the Micronaut Doc's does not have a concrete example in the use of SecurityRule & Annotation to validate the JWT.

Thanks.

**Comment by adrnola** [comment]
@jameskleeh - Thanks. I did some more investigating into this.

I now have the following configuration:

[1] CustomJwtClaimsValidator implements GenericJwtClaimsValidator - which I use for Authentication of JWT (isIssuerValid && isAudienceValid && isGrantTypeValid && isAuthorizedPartyValid && isNotExpired) by validating using application.xml entries below

[2] I created a Custom SecurityRule and created a corresponding Class/Method Annotation - which I use for Authorization (which checks value of the scope).

[3] applciation.xml
  security:
    enabled: true
    token:
      jwt:
        signatures:
          jwks:
            auth0:
              url: https://myApp-test.auth0.com/.well-known/jwks.json
              # These entries are not Micronaut Framework Specific - They are Project Specific.
              # They are placed here as they are JWT/Auth0 specific.
              issuer: https://myApp-test.eu.auth0.com/
              audience: https://myApp-test.com
              grant-type: client-credentials

which worked a treat and is very flexible in my Controller:
@JwtAuthorization(scope = "foo:bar")
@Post(value = "{something}")
public HttpResponse<AppResponse> doSomething()

I am just wondering what is best practice here, move Authentication from CustomJwtClaimsValidator to SecurityRule and use SecurityRule to do both Authentication & Authorization ?

I think where mine and other people (on this thread) confusion arouse is that the Micronaut Doc's does not have a concrete example in the use of SecurityRule & Annotation to validate the JWT.

Thanks.

**Comment by frehov** [comment]
I'd like to chime in here with the same answer I've fiven on #254.

If you're not including roles in the claims, then you can use `micronaut.security.token.roles-name=scope` to remap the ttribute used by the default `RolesFinder` and use the `@RolesAllowed` annotation to achieve securing routes with scopes from JWT.
This would only work as long as you're not dependenent on mapping roles from the default role-attribute as well.

This answer may be skirting the core issue about providing a mechanism to check other claims other than scope, but for those that only need to remap where roles are read from to secure their application, it should be fine.

This property has afaik been available since micronaut 1.0.0
